tools/amplify-preview/amplify.go

```diff
@@ -0,0 +1,338 @@
+*/
+
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/arn"
+	"github.com/aws/aws-sdk-go-v2/service/amplify"
+	"github.com/aws/aws-sdk-go-v2/service/amplify/types"
+)
+
+var (
+	errBranchNotFound           = errors.New("branch not found")
+	errNoJobForBranch           = errors.New("current branch has no jobs")
+	amplifyJobCompletedStatuses = map[types.JobStatus]struct{}{
+		types.JobStatusFailed:    {},
+		types.JobStatusCancelled: {},
+		types.JobStatusSucceed:   {},
+	}
+)
+
+const (
+	logKeyAppID      = "appID"
+	logKeyBranchName = "branchName"
+	logKeyJobID      = "jobID"
+
+	amplifyMarkdownHeader = "Amplify deployment status"
+	amplifyDefaultDomain  = "amplifyapp.com"
+)
+
+type AmplifyPreview struct {
+	appIDs     []string
+	branchName string
+	client     *amplify.Client
+}
+
+type aggregatedError struct {
+	perAppErr map[string]error
+	message   string
+}
+
+func (amp *AmplifyPreview) FindExistingBranch(ctx context.Context) (*types.Branch, error) {
+	type resp struct {
+		appID string
+		data  *amplify.GetBranchOutput
+		err   error
+	}
+	var wg sync.WaitGroup
+	wg.Add(len(amp.appIDs))
+	resultCh := make(chan resp, len(amp.appIDs))
+
+	for _, appID := range amp.appIDs {
+		go func() {
+			defer wg.Done()
+			branch, err := amp.client.GetBranch(ctx, &amplify.GetBranchInput{
+				AppId:      aws.String(appID),
+				BranchName: aws.String(amp.branchName),
+			})
+			resultCh <- resp{
+				appID: appID,
+				data:  branch,
+				err:   err,
+			}
+		}()
+	}
+
+	wg.Wait()
+	close(resultCh)
+
+	failedResp := aggregatedError{
+		perAppErr: map[string]error{},
+		message:   "failed to fetch branch",
+	}
+
+	for resp := range resultCh {
+		var errNotFound *types.NotFoundException
+		if errors.As(resp.err, &errNotFound) {
+			logger.Debug("Branch not found", logKeyAppID, resp.appID, logKeyBranchName, amp.branchName)
+			continue
+		} else if resp.err != nil {
+			failedResp.perAppErr[resp.appID] = resp.err
+			continue
+		}
+
+		if resp.data != nil {
+			return resp.data.Branch, nil
+		}
+	}
+
+	if err := failedResp.Error(); err != nil {
+		return nil, err
+	}
+
+	return nil, errBranchNotFound
+}
+
+func (amp *AmplifyPreview) CreateBranch(ctx context.Context) (*types.Branch, error) {
+	failedResp := aggregatedError{
+		perAppErr: map[string]error{},
+		message:   "failed to create branch",
+	}
+
+	for _, appID := range amp.appIDs {
+		resp, err := amp.client.CreateBranch(ctx, &amplify.CreateBranchInput{
+			AppId:           aws.String(appID),
+			BranchName:      aws.String(amp.branchName),
+			Description:     aws.String("Branch generated for PR TODO"),
+			Stage:           types.StagePullRequest,
+			EnableAutoBuild: aws.Bool(true),
+		})
+
+		var errLimitExceeded *types.LimitExceededException
+		if errors.As(err, &errLimitExceeded) {
+			logger.Debug("Reached branches limit", logKeyAppID, appID)
```

Comment:
Reaching limit is expected behavior in this case.
Because amplify has limit of 50 branches per app, if first app is full we want to move to the next one where there still room left.